archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,718 @@
+      (a - d) - 1 / (2 : ℝ) * (b - e) - (c - f),
+    by {intros, ring_nf},
+    rw [log_stirling_seq_formula, log_stirling_seq_formula, h_reorder],
+    repeat {rw [log_div, factorial_succ]},
+    push_cast,
+    repeat {rw log_mul},
+    rw log_exp,
+    ring_nf,
+    all_goals {norm_cast},
+    all_goals {try {refine mul_ne_zero _ _}, try {exact succ_ne_zero _}},
+    any_goals {exact factorial_ne_zero m},
+    any_goals {exact exp_ne_zero 1},
+    simp },
+  { apply_instance }
+end
+
+/-- The sequence `log_stirling_seq ∘ succ` is monotone decreasing -/
+lemma log_stirling_seq'_antitone : antitone (log_stirling_seq ∘ succ) :=
+begin
+  apply antitone_nat_of_succ_le,
+  intro n,
+  refine sub_nonneg.mp _,
+  rw ← succ_eq_add_one,
+  refine has_sum.nonneg _ (log_stirling_seq_diff_has_sum n),
+  norm_num,
+  simp only [one_div],
+  intro m,
+  refine mul_nonneg _ _,
+  all_goals {refine inv_nonneg.mpr _, norm_cast, exact (zero_le _)},
+end
+
+/--
+We have the bound  `log_stirling_seq n - log_stirling_seq (n+1) ≤ 1/(2n+1)^2* 1/(1-(1/2n+1)^2)`.
+-/
+lemma log_stirling_seq_diff_le_geo_sum : ∀ (n : ℕ),
+  log_stirling_seq n.succ - log_stirling_seq n.succ.succ ≤
+  (1 / (2 * n.succ + 1)) ^ 2 / (1 - (1 / (2 * n.succ + 1)) ^ 2) :=
+begin
+  intro n,
+  have h_nonneg : 0 ≤ ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2),
+  by { rw [cast_succ, one_div, inv_pow, inv_nonneg], norm_cast, exact zero_le', },
+  have g : has_sum (λ (k : ℕ), ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) ^ k.succ)
+    ((1 / (2 * n.succ + 1)) ^ 2 / (1 - (1 / (2 * n.succ + 1)) ^ 2)) :=
+  begin
+    have h_pow_succ := λ (k : ℕ),
+      symm (pow_succ ((1 / (2 * ((n : ℝ) + 1) + 1)) ^ 2) k),
+    have hlt : ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) < 1, by
+    { simp only [cast_succ, one_div, inv_pow],
+      refine inv_lt_one _,
+      norm_cast,
+      simp only [nat.one_lt_pow_iff, ne.def, zero_eq_bit0, nat.one_ne_zero, not_false_iff,
+        lt_add_iff_pos_left, canonically_ordered_comm_semiring.mul_pos, succ_pos', and_self], },
+    exact (has_sum_geometric_of_lt_1 h_nonneg hlt).mul_left ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2)
+  end,
+  have hab :
+    ∀ (k : ℕ), (1 / (2 * (k.succ : ℝ) + 1)) * ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) ^ k.succ ≤
+    ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) ^ k.succ :=
+  begin
+    intro k,
+    have h_zero_le : 0 ≤ ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) ^ k.succ := pow_nonneg h_nonneg _,
+    have h_left : 1 / (2 * (k.succ : ℝ) + 1) ≤ 1, by
+    { simp only [cast_succ, one_div],
+      refine inv_le_one _,
+      norm_cast,
+      exact (le_add_iff_nonneg_left 1).mpr zero_le', },
+    exact mul_le_of_le_one_left h_zero_le h_left,
+  end,
+  exact has_sum_le hab (log_stirling_seq_diff_has_sum n) g,
+end
+
+/--
+We have the bound  `log_stirling_seq n - log_stirling_seq (n+1)` ≤ 1/(4n(n+1))
+-/
+lemma log_stirling_seq_sub_log_stirling_seq_succ (n : ℕ) :
+  log_stirling_seq n.succ - log_stirling_seq n.succ.succ ≤ 1 / (4 * n.succ * n.succ.succ) :=
```

Comment:
A general principle is to claim only what you need, as early as possible.  This saves effort later.  So here I'd switch the RHS of the lemma statement to ```1 / (4 * n.succ ^ 2)```.